[SPARK-20456][Docs] Add examples for functions collection for pyspark

map222 opened the pull request:
## What changes were proposed in this pull request?

This adds documentation to many functions in pyspark.sql.functions.py:
`upper`, `lower`, `reverse`, `unix_timestamp`, `from_unixtime`, `rand`, `randn`, `collect_list`, `collect_set`, `lit`
Add units to the trigonometry functions.
Renames columns in datetime examples to be more informative.
Adds links between some functions.

## How was this patch tested?

`./dev/lint-python`
`python python/pyspark/sql/functions.py`
`./python/run-tests.py --module pyspark-sql`

map222 committed:
First revision: trigonometry units, lit, collect_set, collect_list, unix_timestamp, from_unixtime

map222 commented:
@HyukjinKwon I ended up not making examples for the aggregate functions, as I didn't make a good dataframe to demonstrate them. I could add more examples for the string functions if you think that is a good idea. There are dozens of functions that could be documented, I'm not sure how far we want to go, or which ones need it.

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #76477 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76477/testReport)** for PR 17865 at commit [`91515c6`](https://github.com/apache/spark/commit/91515c620287e193c6d208038025fe194740e4d2).

SparkQA commented:
**[Test build #76477 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76477/testReport)** for PR 17865 at commit [`91515c6`](https://github.com/apache/spark/commit/91515c620287e193c6d208038025fe194740e4d2).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76477/
Test FAILed.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -153,7 +173,7 @@ def _():
 _binary_mathfunctions = {
     'atan2': 'Returns the angle theta from the conversion of rectangular coordinates (x, y) to' +
-             'polar coordinates (r, theta).',
+             'polar coordinates (r, theta). Units in radians.',
```
I am unclear we should note this for every instance and users really gets confused as I see these use Scala/Java's built-in library. I wonder if there is an example that supports this differently in other libraries?

map222 replied:
Most libraries seem to default to radians. However, I checked the R, numpy, and MATLAB notes for common trigonometry functions, and they all note the units in the function documentation, e.g.:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.sin.html
https://www.mathworks.com/help/matlab/ref/sin.html

HyukjinKwon replied:
I see. What do you think about adding this in `:param`?

map222 replied:
I added `:param:` to both `acos` and `asin` using two methods. What do you think?

map222 replied:
@HyukjinKwon Any opinion on how to do the `:param:` for the units?

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -206,17 +226,20 @@ def _():
 def approxCountDistinct(col, rsd=None):
     """
-    .. note:: Deprecated in 2.1, use approx_count_distinct instead.
+    .. note:: Deprecated in 2.1, use :func:`approx_count_distinct instead`.
```
Probably `` :func:`approx_count_distinct` ``?

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -397,7 +420,7 @@ def input_file_name():
 @since(1.6)
 def isnan(col):
-    """An expression that returns true iff the column is NaN.
+    """An expression that returns true if the column is NaN.
```
I think "iff" is the abbreviation for "if and only if". I don't think it is worth changing.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -456,7 +479,7 @@ def monotonically_increasing_id():
     """Returns col1 if it is not NaN, or col2 if col1 is NaN.
 
-    Both inputs should be floating point columns (DoubleType or FloatType).
+    Both inputs should be floating point columns (:class:`DoubleType` or FloatType).
```
I think we should link both `DoubleType` and `FloatType ` all or not.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1120,12 +1159,12 @@ def from_utc_timestamp(timestamp, tz):
     """
-    Given a timestamp, which corresponds to a certain time of day in the given timezone, returns
-    another timestamp that corresponds to the same time of day in UTC.
+    Given a `timestamp`, which corresponds to a time of day in the timezone `tz`,
```
Should this be ``` ``timestamp`` ``` not `` `timestamp` ``?

map222 replied:
I had a quick question about this. The documentation seems pretty inconsistent about when to use \` vs \`\`. For example, in [approx_count_distinct](http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.functions.approx_count_distinct), the parameter `col` is referred to with ``, while in [rpad](http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html#pyspark.sql.functions.rpad), the parameters are referred to with `.

Numpy appears to use single backticks for parameters:
https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections

Is there a reference for this?

HyukjinKwon replied:
No, I think we have a rule about this up to my knowledge. Thank you for the pointers and looking into this. Let's follow the majority then for now.

HyukjinKwon replied:
No, I don't think we have a rule about this up to my knowledge. Thank you for the pointers and looking into this. Let's follow the majority then for now.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -67,9 +67,16 @@ def _():
 
+_lit_doc = """
+    Creates a :class:`Column` of literal value. Supports basic types like :class:`IntegerType`,
+    :class:`FloatType`, :class:`BooleanType`, and :class:`StringType`
```
I would like to keep this identical with the one in `functions.scala` to reduce overhead when someone sweeps the same documentation changes across APIs in other languages.

If the additional informations are Python-specific, let's add it in `::note`.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -793,8 +824,8 @@ def date_format(date, format):
 
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['a'])
-    >>> df.select(date_format('a', 'MM/dd/yyy').alias('date')).collect()
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['time'])
```
Okay. I guess it is a documentation improvement to use a bit more meaningful name over an arbitrary name `a`. Let's match these to existing names such as `ts` or `t` (abbreviation for timestamp) or `dt` (abbreviation for `datetime.datetime`).

map222 replied:
I am replacing all of the timestamps of form `yyyy-MM-dd` with `dt`, and all of the `yyyy-MM-dd HH:mm:ss` with ts, with some exceptions. Hopefully this addresses this point.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -910,8 +941,8 @@ def weekofyear(col):
 
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['a'])
-    >>> df.select(weekofyear(df.a).alias('week')).collect()
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['time'])
```
Let's use `d` for `DateType` or `datetime.date` similarly with other existing names.

HyukjinKwon reviewed (commented):
I have not finished my review. I will take a look again.
Let's fix up lint check failures.

In general, I think I am not sure of changing the names to `time`. It does not look particularly better to me but unnecessary changes.

Also, let's check out the contents in Scala/R ones. I think we should keep them identical in most cases.

If there are Python-specific contents, let's use `::note`. If it is R-specific, ` Note: `. If it is Scala/Java specific, `@note`.

HyukjinKwon commented:
(Thank you @gatorsmile for triggering the test)

gatorsmile commented on python/pyspark/sql/functions.py:
```diff
@@ -409,7 +432,7 @@ def isnan(col):
 @since(1.6)
 def isnull(col):
-    """An expression that returns true iff the column is null.
+    """An expression that returns true if the column is null.
```
`the column`? This is misleading. We should make the Python documents consistent with what we did in Scala. 
For example, `isNull` in Scala APIs is described as
> Returns true if `expr` is null, or false otherwise.

Ref: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala#L280

gatorsmile commented:
Could you please check the documents we did in Scala APIs? It sounds like we forgot to update the Python function descriptions when we did the change in the Scala APIs.

map222 commented:
@gatorsmile I checked four functions, `approx_count_distinct`, `coalesce`, `covar_samp`, and `countDistinct`, comparing the python and Scala documentation. None of them are the same. My guess is that the python docs differ for most functions.

map222 committed:
Changed names of time columns, removed extra words for lit, fixed some typos

SparkQA commented:
**[Test build #76625 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76625/testReport)** for PR 17865 at commit [`ca8b5f7`](https://github.com/apache/spark/commit/ca8b5f7d666bd13a515ba1358e4f69ff13df9711).

SparkQA commented:
**[Test build #76625 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76625/testReport)** for PR 17865 at commit [`ca8b5f7`](https://github.com/apache/spark/commit/ca8b5f7d666bd13a515ba1358e4f69ff13df9711).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76625/
Test FAILed.

gatorsmile commented:
@map222 Unfortunately, our PySpark did not follow what we did in Scala. Will review it more carefully in the future. Thanks!

map222 committed:
Changed backticks. Tested new :param: for angle functions

map222 commented:
Most recent commit reverts the "if" to "iff", changes all the backticks for column names to single backtick, and tried new `:param:` option for angle columns.

SparkQA commented:
**[Test build #76763 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76763/testReport)** for PR 17865 at commit [`dd7a397`](https://github.com/apache/spark/commit/dd7a3971275aae97f9adfd82166df23b038eb950).

SparkQA commented:
**[Test build #76763 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76763/testReport)** for PR 17865 at commit [`dd7a397`](https://github.com/apache/spark/commit/dd7a3971275aae97f9adfd82166df23b038eb950).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76763/
Test PASSed.

map222 commented:
Could I get a review for this? I think the only remaining questions is whether (and how) to note the units for the trigonometry functions, like here:
https://github.com/map222/spark/blob/dd7a3971275aae97f9adfd82166df23b038eb950/python/pyspark/sql/functions.py#L100-L105

map222 commented:
@HyukjinKwon @gatorsmile bump

gatorsmile commented:
Also cc @ueshin

HyukjinKwon commented:
Will take a look to help too soon. Sorry for the delay.

ueshin reviewed (commented):
I added some minor style-related comments.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -189,15 +210,15 @@ def _():
 
 for _name, _doc in _functions.items():
-    globals()[_name] = since(1.3)(_create_function(_name, _doc))
+    globals()[_name] = since(1.3)(ignore_unicode_prefix(_create_function(_name, _doc)))
```
Hm, this then applies `ignore_unicode_prefix` to all docs ... This looks only added for `_lit_doc`. Could we simply avoid the unicode example in the lit doc rather than applying this to all function docs?

map222 replied:
`ignore_unicode_prefix` is necessary to get the tests to pass both python 2 and python 3. I had problems without it on my previous documentation PR: https://github.com/apache/spark/pull/17469#issuecomment-295896057

HyukjinKwon replied:
Yea, I didn't mean to not use but avoid applying all other docs (more specifically the ones in `_functions`).  This basically replaces u prefix via regex and rather a workaround IMHO. So, I think we should avoid applying this to other doctests not using unicode.

HyukjinKwon replied:
And .. the easiest way I think is to take out the examples or not use unicode in the doctests for now as `ignore_unicode_prefix` requires a function.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -189,15 +210,15 @@ def _():
     globals()[_name] = since(1.6)(_create_window_function(_name, _doc))
 for _name, _doc in _functions_1_6.items():
-    globals()[_name] = since(1.6)(_create_function(_name, _doc))
+    globals()[_name] = since(1.6)(ignore_unicode_prefix(_create_function(_name, _doc)))
```
This one looks also the same instance. Let's avoid adding unicode examples in `_collect_list_doc` and `_collect_set_doc` for now if we are all fine.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -206,17 +227,20 @@ def _():
 @since(2.1)
 def approx_count_distinct(col, rsd=None):
-    """Returns a new :class:`Column` for approximate distinct count of ``col``.
+    """Returns a new :class:`Column` for approximate distinct count of column `col`.
```
Sounds like it also needs leading `Aggregate function: ` doc - https://github.com/apache/spark/blob/3a840048ed3501e06260b7c5df18cc0bbdb1505c/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L225.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -206,17 +227,20 @@ def _():
 
-    >>> df.agg(approx_count_distinct(df.age).alias('c')).collect()
-    [Row(c=2)]
+    :param rsd: Residual (float). The approximate count will be within this fraction of the true
```
For this one, let's add the contents in Scala/R one too - https://github.com/apache/spark/blob/3a840048ed3501e06260b7c5df18cc0bbdb1505c/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L245 and https://github.com/apache/spark/blob/7f203a248f94df6183a4bc4642a3d873171fef29/R/pkg/R/functions.R#L2243

This one in Scala/R look already having the param contents to match with.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -793,9 +824,9 @@ def date_format(date, format):
-    >>> df.select(date_format('a', 'MM/dd/yyy').alias('date')).collect()
-    [Row(date=u'04/08/2015')]
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['dt'])
+    >>> df.select(date_format('dt', 'MM/dd/yyy').alias('dt2')).collect()
```
I think `date` was also fine.

HyukjinKwon replied:
`date` to `dt2` sounds unnecessary change.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -962,9 +993,9 @@ def add_months(start, months):
-    >>> df = spark.createDataFrame([('2015-04-08',)], ['d'])
-    >>> df.select(add_months(df.d, 1).alias('d')).collect()
-    [Row(d=datetime.date(2015, 5, 8))]
+    >>> df = spark.createDataFrame([('2015-04-08',)], ['dt'])
```
I think changing arbitrary name to `dt` or `ts` is fine but not worth changing `t`, `timestamp`, `d` or `date`. These are already readable.

map222 replied:
In the original example, the column name `d` is reused for the new column. I think using a new column name helps differentiate the original date from the new date.

HyukjinKwon replied:
Okay, but let's sweep the other instances here if there are.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1254,23 +1294,41 @@ def hash(*cols):
 
 for _name, _doc in _string_functions.items():
-    globals()[_name] = since(1.5)(_create_function(_name, _doc))
+    globals()[_name] = since(1.5)(ignore_unicode_prefix(_create_function(_name, _doc)))
```
Hm.. can we avoid applying this to all docs too? One way is just to take out the examples.  These string function examples looks easy and straightforward.

map222 replied:
I think it is helpful to be clear on how `lower` and `upper` should be used. This week I tried to use `lower` as a column function, rather than as an outside function. I can remove the docstrings if you think avoiding `ignore_unicode_prefix` is important.

HyukjinKwon replied:
In that way, we should add all examples for each function here. I am not too sure if we need them all for now. At least per - https://issues.apache.org/jira/browse/SPARK-17963?focusedCommentId=15581022&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15581022.

HyukjinKwon reviewed (commented):
I also left some comments.

map222 committed:
Addressed nits. Fixed rsd. Fixed units on trig functions.

map222 commented on python/pyspark/sql/functions.py:
```diff
@@ -92,14 +98,16 @@ def _():
+    'asin': """Computes the sine inverse of the given value; the returned angle is in the range
+            -pi/2 through pi/2.
+
+            :param col: float column, units in radians.""",
```
I was trying two different ways to see which worked best. I now use the `:param col:` version

SparkQA commented:
**[Test build #78029 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78029/testReport)** for PR 17865 at commit [`a61173f`](https://github.com/apache/spark/commit/a61173f2172cc716c76abf2e14ca21ff77ab08b2).

SparkQA commented:
**[Test build #78029 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78029/testReport)** for PR 17865 at commit [`a61173f`](https://github.com/apache/spark/commit/a61173f2172cc716c76abf2e14ca21ff77ab08b2).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78029/
Test FAILed.

SparkQA commented:
**[Test build #78031 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78031/testReport)** for PR 17865 at commit [`8c34c8b`](https://github.com/apache/spark/commit/8c34c8b9073863178ff74c18a01dcf0db32e6805).

SparkQA commented:
**[Test build #78031 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78031/testReport)** for PR 17865 at commit [`8c34c8b`](https://github.com/apache/spark/commit/8c34c8b9073863178ff74c18a01dcf0db32e6805).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78031/
Test FAILed.

holdenk commented:
Thanks for working on this :) More complete documentation is certainly useful. Looking at the "How was this patch tested?" section you might also want to build the docs (there are instructions in `docs/README.md`) to see how stuff is formatted.

map222 committed:
Removed string docs, removed ignore_unicode_prefix where unnecessary, changed date column names back, and fixed the doc warning.

SparkQA commented:
**[Test build #78158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78158/testReport)** for PR 17865 at commit [`29bcb67`](https://github.com/apache/spark/commit/29bcb67eb85bf79781d3a3bf1cb35cdcfe433f5c).

map222 committed:
Fixed the iff

SparkQA commented:
**[Test build #78159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78159/testReport)** for PR 17865 at commit [`b4c04e7`](https://github.com/apache/spark/commit/b4c04e762e122f575292af7aa00934bb60501aa2).

map222 commented:
Ok, I removed the `ignore_unicode_prefix`, and it didn't have any test problems. Also removed the date column renames where the original name was ok. Finally, removed the string function documentation.

SparkQA commented:
**[Test build #78158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78158/testReport)** for PR 17865 at commit [`29bcb67`](https://github.com/apache/spark/commit/29bcb67eb85bf79781d3a3bf1cb35cdcfe433f5c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78158/
Test FAILed.

SparkQA commented:
**[Test build #78159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78159/testReport)** for PR 17865 at commit [`b4c04e7`](https://github.com/apache/spark/commit/b4c04e762e122f575292af7aa00934bb60501aa2).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78159/
Test FAILed.

map222 committed:
Fixed unicode

SparkQA commented:
**[Test build #78183 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78183/testReport)** for PR 17865 at commit [`89138ea`](https://github.com/apache/spark/commit/89138ea41bf27c8dd6549ef6c38395f3ccf6d843).

SparkQA commented:
**[Test build #78183 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78183/testReport)** for PR 17865 at commit [`89138ea`](https://github.com/apache/spark/commit/89138ea41bf27c8dd6549ef6c38395f3ccf6d843).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78183/
Test FAILed.

map222 committed:
Re-included the ignore_unicode_prefix

SparkQA commented:
**[Test build #78188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78188/testReport)** for PR 17865 at commit [`e1bb723`](https://github.com/apache/spark/commit/e1bb723a04900704889fa0c66a14723193898bf0).

map222 committed:
Trying to undo merge

SparkQA commented:
**[Test build #78189 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78189/testReport)** for PR 17865 at commit [`e6cb8a1`](https://github.com/apache/spark/commit/e6cb8a1e830ffa18d6bdbcbeb2aa6456bc6569da).

SparkQA commented:
**[Test build #78188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78188/testReport)** for PR 17865 at commit [`e1bb723`](https://github.com/apache/spark/commit/e1bb723a04900704889fa0c66a14723193898bf0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78188/
Test PASSed.

map222 commented:
@HyukjinKwon I could not get it to pass the tests without the unicode and `ignore_unicode_prefix`

SparkQA commented:
**[Test build #78189 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78189/testReport)** for PR 17865 at commit [`e6cb8a1`](https://github.com/apache/spark/commit/e6cb8a1e830ffa18d6bdbcbeb2aa6456bc6569da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78189/
Test PASSed.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -147,7 +173,7 @@ def _():
 _binary_mathfunctions = {
     'atan2': 'Returns the angle theta from the conversion of rectangular coordinates (x, y) to' +
-             'polar coordinates (r, theta).',
+             'polar coordinates (r, theta). Units in radians.',
```
I think we should ...

1. copy and paste `Units in radians.` to both https://github.com/apache/spark/blob/e5387018e76a9af1318e78c4133ee68232e6a159/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L1342 and https://github.com/apache/spark/blob/8965fe764a4218d944938aa4828072f1ad9dbda7/R/pkg/R/functions.R#L2030 just to be consistent

2. Use Python specific notation to describe this here for now.

3. Just revert this change for now.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1258,7 +1303,7 @@ def hash(*cols):
     'lower': 'Converts a string column to lower case.',
     'upper': 'Converts a string column to upper case.',
-    'reverse': 'Reverses the string column and returns it as a new string column.',
+    'reverse': 'Reverses a string column and returns it as a new string column.',
```
ditto. I guess we should match this change to R/Scala ones too.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -109,15 +118,33 @@ def _():
+    Aggregate function: returns a list of objects with duplicates.
+
+    >>> df2 = spark.createDataFrame([('Alice', 2), ('Bob', 5), ('Alice', 99)], ('name', 'age'))
+    >>> df2.agg(collect_list('name')).collect()
```
I think we can avoid `ignore_unicode_prefix` if we call this function with `age` for this and the one below.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -67,9 +67,15 @@ def _():
 
+    >>> df.withColumn('height', lit(5)).withColumn('spark_user', lit(True)).collect()
+    [Row(age=2, name=u'Alice', height=5, spark_user=True),
+     Row(age=5, name=u'Bob', height=5, spark_user=True)]
```
Here too, I think we can just do something like `df.select(hight, lit(5)).collect()` or `show()` to avoid `ignore_unicode_prefix`.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -200,17 +226,20 @@ def _():
 @since(2.1)
-def approx_count_distinct(col, rsd=None):
-    """Returns a new :class:`Column` for approximate distinct count of ``col``.
+def approx_count_distinct(col, rsd=0.05):
```
Why do we change this to `0.05`? We then should change the default values in both places in `HyperLogLogPlusPlus` and  here if we happen to change this in any event. Let's leave it `None`.

HyukjinKwon replied:
Okay, I double checked some APIs have them but let's leave this out. It sounds requiring fixing/checking logic below and the deprecated one above. I assume this should be doc only change PR.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -267,8 +296,7 @@ def coalesce(*cols):
 def corr(col1, col2):
-    """Returns a new :class:`Column` for the Pearson Correlation Coefficient for ``col1``
-    and ``col2``.
+    """Returns a new :class:`Column` for the Pearson Correlation Coefficient for `col1` and `col2`.
```
Honestly, this is rather unnecessary change. Both before and after are fine and do not affect the documentation rendering.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -969,8 +1005,8 @@ def months_between(date1, date2):
 
-    >>> df = spark.createDataFrame([('1997-02-28 10:30:00', '1996-10-30')], ['t', 'd'])
-    >>> df.select(months_between(df.t, df.d).alias('months')).collect()
+    >>> df = spark.createDataFrame([('1997-02-28 10:30:00', '1996-10-30')], ['date1', 'date2'])
```
`t` and `d` looks already fine.

map222 replied:
I reverted all the other ones, but I like this one because `date1` and `date2` match the docstring. Is that reasonable?

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1073,12 +1109,17 @@ def last_day(date):
     format.
+
+    >>> time_df = spark.createDataFrame([(1428476400, )], ['unix_time'])
+    >>> time_df.select(from_unixtime('unix_time').alias('ts') ).collect()
```
nit: `) )` -> `))`.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1105,9 +1150,9 @@ def from_utc_timestamp(timestamp, tz):
 
-    >>> df = spark.createDataFrame([('1997-02-28 10:30:00',)], ['t'])
-    >>> df.select(from_utc_timestamp(df.t, "PST").alias('t')).collect()
-    [Row(t=datetime.datetime(1997, 2, 28, 2, 30))]
```
`t` to `ts` looks unnecessary change here and the one below.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -312,7 +338,7 @@ def covar_samp(col1, col2):
 @since(1.3)
 def countDistinct(col, *cols):
-    """Returns a new :class:`Column` for distinct count of ``col`` or ``cols``.
+    """Returns a new :class:`Column` for distinct count of `col` or `cols`.
```
I guess we did not decide which one of `` `...` `` or ``` ``...`` ``` is more preferred and right. Let's avoid sweeping them. Adding new one is fine (and I usually follow majority in this case) but changing the existing one might be not if we don't know which one is not preferred. We might happen to change it forward and backward in time.

HyukjinKwon reviewed (commented):
I believe these are all from me.

map222 committed:
Switched `lit`, `collect_set`, and `collect_list` to numeric examples. Removed `ignore_unicode_prefix` from those docs. Added units to Scala and R. Reverted datetime name changes

map222 commented:
@HyukjinKwon I think I addressed all of your comments. Thank you for your detailed review!

SparkQA commented:
**[Test build #78586 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78586/testReport)** for PR 17865 at commit [`9e1e9e1`](https://github.com/apache/spark/commit/9e1e9e19bcfaf0a3fb5bf3264058b46573877b76).

SparkQA commented:
**[Test build #78586 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78586/testReport)** for PR 17865 at commit [`9e1e9e1`](https://github.com/apache/spark/commit/9e1e9e19bcfaf0a3fb5bf3264058b46573877b76).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78586/
Test FAILed.

felixcheung commented on R/pkg/R/functions.R:
```diff
@@ -426,7 +426,7 @@ setMethod("covar_pop", signature(col1 = "characterOrColumn", col2 = "characterOr
 #' cos
 #'
-#' Computes the cosine of the given value.
+#' Computes the cosine of the given value. Units in radians.
```
looks like this will conflict with #18371

HyukjinKwon replied:
Yes, if that one is merged first, we should resolve the conflict here. If this one gets merged first, that one should resolve the conflict. Describing the same information across APIs looks reasonable to me.

HyukjinKwon replied:
BTW, just while you are here @felixcheung, does this change itself look okay to you?

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -200,17 +225,20 @@ def _():
 @since(2.1)
 def approx_count_distinct(col, rsd=None):
-    """Returns a new :class:`Column` for approximate distinct count of ``col``.
+    """Aggregate function. Returns a new :class:`Column` for approximate distinct count of column `col`.
```
little nit `. R` -> `: r`.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -282,8 +309,7 @@ def corr(col1, col2):
 def covar_pop(col1, col2):
-    """Returns a new :class:`Column` for the population covariance of ``col1``
-    and ``col2``.
+    """Returns a new :class:`Column` for the population covariance of `col1` and `col2`.
```
little nit: `` `col1` `` -> ``` ``col1`` ``` and the same one below.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1116,12 +1160,12 @@ def from_utc_timestamp(timestamp, tz):
 @since(1.5)
 def to_utc_timestamp(timestamp, tz):
     """
-    Given a timestamp, which corresponds to a certain time of day in the given timezone, returns
```
Let's revert this change or do this the same thing to Scala/R too if it matters.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -1258,7 +1302,7 @@ def hash(*cols):
     'lower': 'Converts a string column to lower case.',
     'upper': 'Converts a string column to upper case.',
-    'reverse': 'Reverses the string column and returns it as a new string column.',
+    'reverse': 'Reverses a string column and returns it as a new string column.',
```
ditto. I guess we should revert this change or match this change to R/Scala ones too if it matters.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -473,10 +503,15 @@ def rand(seed=None):
     """Generates a column with independent and identically distributed (i.i.d.) samples from
     the standard normal distribution.
+
+    >>> df.withColumn('randn', randn(seed=42) ).collect()
```
little nit: `) )` -> `))`

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -95,10 +100,13 @@ def _():
             '-pi/2 through pi/2.',
-    'atan': 'Computes the tangent inverse of the given value.',
+    'atan': 'Computes the tangent inverse of the given value; the returned angle is in the range' +
+            '-pi/2 through pi/2',
```
Let's revert this change or add the same information in Scala/R too.

HyukjinKwon reviewed (commented):
It looks good to me except for few minor comments and the comment above ^.

ueshin commented on python/pyspark/sql/functions.py:
```diff
@@ -1073,12 +1108,17 @@ def last_day(date):
     representing the timestamp of that moment in the current system time zone in the given
     format.
+
+    >>> time_df = spark.createDataFrame([(1428476400, )], ['unix_time'])
```
little nit: `(1428476400, )` -> `(1428476400,)`

ueshin reviewed (commented):
LGTM except for a minor comment and comments above.

map222 committed:
Added atan range to scala and R, fixed nits

SparkQA commented:
**[Test build #79118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79118/testReport)** for PR 17865 at commit [`f17f332`](https://github.com/apache/spark/commit/f17f332dd97b948f8dd31eb2b18c1e11dc7fead0).

map222 commented:
I think I addressed everything.

ueshin commented:
Looks like there is a merge conflict. Can you fix it?

map222 commented:
Do I just have to `git rebase` on master?

ueshin commented:
Yes, or you can merge master and fix the conflict.

SparkQA commented:
**[Test build #79118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79118/testReport)** for PR 17865 at commit [`f17f332`](https://github.com/apache/spark/commit/f17f332dd97b948f8dd31eb2b18c1e11dc7fead0).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79118/
Test PASSed.

felixcheung commented on R/pkg/R/functions.R:
```diff
@@ -148,7 +148,8 @@ setMethod("asin",
 #'
-#' Computes the tangent inverse of the given value.
+#' Computes the tangent inverse of the given value; the returned angle is in the range
+#' -pi/2 through pi/2
```
this has been moved in master - you should see when you rebase

SparkQA commented:
**[Test build #79162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79162/testReport)** for PR 17865 at commit [`f5d0d0f`](https://github.com/apache/spark/commit/f5d0d0f2d1649305213e50d9e76180f449027a4a).

SparkQA commented:
**[Test build #79162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79162/testReport)** for PR 17865 at commit [`f5d0d0f`](https://github.com/apache/spark/commit/f5d0d0f2d1649305213e50d9e76180f449027a4a).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79162/
Test FAILed.

SparkQA commented:
**[Test build #79163 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79163/testReport)** for PR 17865 at commit [`e7935d1`](https://github.com/apache/spark/commit/e7935d15f0728af6ecd74373f6c71d2239360815).

SparkQA commented:
**[Test build #79163 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79163/testReport)** for PR 17865 at commit [`e7935d1`](https://github.com/apache/spark/commit/e7935d15f0728af6ecd74373f6c71d2239360815).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79163/
Test FAILed.

map222 committed:
Hopefully fixed the merge in functions.R

map222 committed:
Fixed whitespace

SparkQA commented:
**[Test build #79164 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79164/testReport)** for PR 17865 at commit [`2c9898b`](https://github.com/apache/spark/commit/2c9898bff7f4b65a303ac2f679cd390700b531ef).

SparkQA commented:
**[Test build #79164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79164/testReport)** for PR 17865 at commit [`2c9898b`](https://github.com/apache/spark/commit/2c9898bff7f4b65a303ac2f679cd390700b531ef).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79164/
Test FAILed.

map222 committed:
Fixed more

SparkQA commented:
**[Test build #79165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79165/testReport)** for PR 17865 at commit [`60af595`](https://github.com/apache/spark/commit/60af5959a340068ea82d4e96c92b634959494f93).

SparkQA commented:
**[Test build #79165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79165/testReport)** for PR 17865 at commit [`60af595`](https://github.com/apache/spark/commit/60af5959a340068ea82d4e96c92b634959494f93).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79165/
Test PASSed.

HyukjinKwon commented on R/pkg/R/functions.R:
```diff
@@ -1900,8 +1901,8 @@ setMethod("year",
-#' (x, y) to polar coordinates (r, theta).
-#'
+#' (x, y) to polar coordinates (r, theta). Units in radians.
+#
```
Sorry... missing `'`...

HyukjinKwon reviewed (approved):
It looks https://github.com/apache/spark/pull/17865/files#r123964695 is missed .. LGTM too.

map222 committed:
Fixed nits

SparkQA commented:
**[Test build #79189 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79189/testReport)** for PR 17865 at commit [`9cc34c8`](https://github.com/apache/spark/commit/9cc34c895eaad5ba4366308a0324c8adb2f9510e).

SparkQA commented:
**[Test build #79189 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79189/testReport)** for PR 17865 at commit [`9cc34c8`](https://github.com/apache/spark/commit/9cc34c895eaad5ba4366308a0324c8adb2f9510e).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79189/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #79207 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79207/testReport)** for PR 17865 at commit [`9cc34c8`](https://github.com/apache/spark/commit/9cc34c895eaad5ba4366308a0324c8adb2f9510e).

SparkQA commented:
**[Test build #79207 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79207/testReport)** for PR 17865 at commit [`9cc34c8`](https://github.com/apache/spark/commit/9cc34c895eaad5ba4366308a0324c8adb2f9510e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79207/
Test PASSed.

felixcheung reviewed (commented):
LGTM. 
how about https://github.com/apache/spark/pull/17865/files#r114929075?

map222 commented:
I switched the examples to numeric examples so that we could avoid using `ignore_unicode_prefix`, as requested by @HyukjinKwon in [this commit](https://github.com/apache/spark/pull/17865/commits/0ab91297c09628377fffc04b3f54216bf2581a89#diff-f5295f69bfbdbf6e161aed54057ea36dL144)

ueshin commented:
How about https://github.com/apache/spark/pull/17865/files#r123964695 ?

HyukjinKwon commented:
Yea. For https://github.com/apache/spark/pull/17865#issuecomment-313303801, just to explain the context (as it is a bit confusing even to me), it was suggested to use `ignore_unicode_prefix` for single instance when the example has an unicode in the doctest _or_ to remove the unicode in the doctest if possible (either should be fine).

I also suggested to not use it for unrelated docstrings - https://github.com/apache/spark/pull/17865#discussion_r121539034.

map222 committed:
Reverted to_utc_timestamp

SparkQA commented:
**[Test build #79343 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79343/testReport)** for PR 17865 at commit [`7dbf35f`](https://github.com/apache/spark/commit/7dbf35f2331d2363191850b3f4e13bf58241bad2).

SparkQA commented:
**[Test build #79343 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79343/testReport)** for PR 17865 at commit [`7dbf35f`](https://github.com/apache/spark/commit/7dbf35f2331d2363191850b3f4e13bf58241bad2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79343/
Test PASSed.

felixcheung reviewed (commented):
any more thought on this?

ueshin commented:
LGTM.

HyukjinKwon commented:
LGTM too.

felixcheung commented:
merged to master. thanks!
